File: JULY-2025/11-07-2025/p2402.kt
class Solution {
    fun mostBooked(n: Int, meetings: Array<IntArray>): Int {
        val sortedMeetings = meetings.sortedBy { it[0] }
        val rooms = Array(n) { Room(it) }
        for (meeting in sortedMeetings) {
            var minRoomAvailable: Room = rooms[0]
            var minAvailabilityTime: Long = Long.MAX_VALUE
            for (room in rooms) {
                if (room.timeToFree <= meeting[0]) {
                    minRoomAvailable = room
                    break
                }
                if (minAvailabilityTime > room.timeToFree) {
                    minAvailabilityTime = room.timeToFree
                    minRoomAvailable = room
                }
            }
            minRoomAvailable.apply {
                timeToFree =
                    if (timeToFree < meeting[0]) meeting[1].toLong()
                    else timeToFree + (meeting[1] - meeting[0])
                noOfMeetings++
            }
            rooms[minRoomAvailable.id] = minRoomAvailable
        }
        return rooms.maxBy { it.noOfMeetings }.id
    }
}

data class Room(
    val id: Int,
    var timeToFree: Long = 0L,
    var noOfMeetings: Int = 0
)

File: JULY-2025/11-07-2025/p2402.scala
object Solution:
  // helper class
  case class MeetingRoom(roomNumber: Int, var lastMeetingEndTime: Int = 0, var meetingsHeld: Int = 0)

  import collection.mutable
  import util.Sorting

  def mostBooked(n: Int, meetings: Array[Array[Int]]): Int =
    Sorting.quickSort(meetings)(Ordering.by(_(0)))
    val rooms = (0 until n).map(MeetingRoom(_))
  
    val freeRooms = mutable.PriorityQueue.from(rooms)((a, b) => b.roomNumber - a.roomNumber)
    val occupiedRooms = mutable.PriorityQueue.empty[MeetingRoom](
      (a, b) =>
        if (a.lastMeetingEndTime == b.lastMeetingEndTime) b.roomNumber - a.roomNumber
        else b.lastMeetingEndTime - a.lastMeetingEndTime
    )
  
    for (Array(start, end) <- meetings)
      // check if rooms can be freed
      while (occupiedRooms.nonEmpty && occupiedRooms.head.lastMeetingEndTime <= start)
        freeRooms.addOne(occupiedRooms.dequeue)
  
      val room = if (freeRooms.nonEmpty) freeRooms.dequeue else occupiedRooms.dequeue
      room.lastMeetingEndTime = math.max(room.lastMeetingEndTime, start) + end - start // start time with possible delay + duration
      room.meetingsHeld += 1
      occupiedRooms.addOne(room)
  
    rooms.maxBy(_.meetingsHeld).roomNumber